core/src/robot/AbstractTransmission.cc

```diff
@@ -0,0 +1,374 @@
+        {
+            PRINT_ERROR("Transmission already attached to a robot. Please 'detach' method before attaching it.");
+            return hresult_t::ERROR_GENERIC;
+        }
+
+        // Make sure the robot still exists
+        if (robot.expired())
+        {
+            PRINT_ERROR("Robot pointer expired or unset.");
+            return hresult_t::ERROR_GENERIC;
+        }
+
+
+        // Copy references to the robot and shared data
+        robot_ = robot;
+
+        // Update the flag
+        isAttached_ = true;
+
+        return hresult_t::SUCCESS;
+    }
+
+    hresult_t AbstractTransmissionBase::detach(void)
+    {
+        if (!isAttached_)
+        {
+            PRINT_ERROR("Transmission not attached to any robot.");
+            return hresult_t::ERROR_GENERIC;
+        }
+
+        // Clear the references to the robot
+        robot_.reset();
+
+        // Unset the Id
+        transmissionIdx_ = -1;
+
+        // Reset constant members
+        jointNames_.clear();
+        jointModelIndices_.clear();
+        jointTypes_.clear();
+        jointPositionIndices_.clear();
+        jointVelocityIndices_.clear();
+        motorNames_.clear();
+        motorIndices_.clear();
+
+        // Update the flags
+        isAttached_ = false;
+        isInitialized_ = false;
+
+        return hresult_t::SUCCESS;
+    }
+
+    hresult_t AbstractTransmissionBase::setOptions(configHolder_t const & /*transmissionOptions*/)
+    {
+        // TODO SetOptions
+        return hresult_t::SUCCESS;
+    }
+
+    configHolder_t AbstractTransmissionBase::getOptions(void) const
+    {
+        return transmissionOptionsHolder_;
+    }
+
+    hresult_t AbstractTransmissionBase::refreshProxies(void)
+    {
+        hresult_t returnCode = hresult_t::SUCCESS;
+
+        if (!isAttached_)
+        {
+            PRINT_ERROR("Transmission not attached to any robot. Impossible to refresh proxies.");
+            returnCode = hresult_t::ERROR_INIT_FAILED;
+        }
+
+        auto robot = robot_.lock();
+        if (returnCode == hresult_t::SUCCESS)
+        {
+            if (!robot)
+            {
+                PRINT_ERROR("Robot has been deleted. Impossible to refresh proxies.");
+                returnCode = hresult_t::ERROR_GENERIC;
+            }
+        }
+
+        if (returnCode == hresult_t::SUCCESS)
+        {
+            if (!isInitialized_)
+            {
+                PRINT_ERROR("Transmission not initialized. Impossible to refresh proxies.");
+                returnCode = hresult_t::ERROR_INIT_FAILED;
+            }
+        }
+
+        if (returnCode == hresult_t::SUCCESS)
+        {
+            if (!robot->getIsInitialized())
+            {
+                PRINT_ERROR("Robot not initialized. Impossible to refresh proxies.");
+                returnCode = hresult_t::ERROR_INIT_FAILED;
+            }
+        }
+
+        jointIndex_t jointModelIdxTmp;
+        for (unsigned int i = 0; i < jointNames_.size(); i++)
```

Comment:
Ok I checked the doc of `std::transform` and now I understand what is going on there. I am not sure if for the readability / clearness of the code this here is such much better than the "simple" loop right now that is understandable at a first glance. I mean I like to learn new things and if there is any advantage of using this I will do so. Just right now I would never come up with a solution like this to solve this task

Reply:
The point is, the standard library is heavily optimized, to do much better than you would do using an explicit loop as you are doing. Using it is a good way to set the standard lib take care of doing all the optimization, which could even improve with new release of C++ without having to do anything yourself. So it is definitely a good practice to replace for loop by std transform and others. But here, the lambda is complicated (it captures stuff) so in such a case, it will probably be impossible for the compiler to optimize it (lambda capturing stuff are completely different beast than basic lambda object internally. Capture lambda are class functors, which basic lambda are orphan function pointers). So I would advance to never use std transform and co if the lambda captures anything.

Reply:
Ok I understand your point, very interesting. thanks for the more detailed explication